### PR TITLE
Improve UpgradeDependencyVersion to update dependency version whose value is defined via a property from the parent POM.

### PR DESCRIPTION
**Pull request by nguyenhoan**
## What's changed?
Improve UpgradeDependencyVersion to update dependency version whose value is defined via a property from the parent POM.

## What's your motivation?
- https://github.com/openrewrite/rewrite/issues/4193

## Anyone you would like to review specifically?
@timtebeek 

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [ ] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

**Comment by nguyenhoan** [comment]
Sure. Thanks!

**Comment by nguyenhoan** [comment]
@timtebeek just wanted to check if you have an ETA for a full review of this PR? Thanks!

**Comment by timtebeek** [comment]
Apologies @nguyenhoan ! Travel & catching up to what I've missed is taking a little longer than I had hoped; I plan to revisit this Monday. At first glance it's quality work; I just want to make sure there's no regressions given how often this recipe is used, and I'll have to update downstream use in tandem. Appreciate your patience & persistence in seeing this through!

**Comment by nguyenhoan** [comment]
Thanks @timtebeek! Please let me know if you need anything from me. Look forward to the release of this enhancement.

**Comment by nguyenhoan** [comment]
@timtebeek just a gentle ping on when this can be merged. Thanks!

File: rewrite-maven/src/main/java/org/openrewrite/maven/UpgradeDependencyVersion.java [hunk 1]
```diff
@@ -125,35 +126,84 @@ public String getDescription() {
...
+            }
+
+            @Override
+            public Xml.Tag visitTag(final Xml.Tag tag, final ExecutionContext executionContext) {
+                ResolvedDependency d = findDependency(tag);
+                if (d != null && d.getRepository() != null) {
+                    // if the resolved dependency exists AND it does not represent an artifact that was parsed
+                    // as a source file, attempt to find a new version.
+                    try {
+                        String newerVersion = findNewerVersion(d.getGroupId(), d.getArtifactId(), d.getVersion(), executionContext);
```
Comment (timtebeek) [inline rewrite-maven/src/main/java/org/openrewrite/maven/UpgradeDependencyVersion.java:153]:
```suggestion
                        String newerVersion = findNewerVersion(d.getGroupId(), d.getArtifactId(), d.getVersion(), ctx);
```

File: rewrite-maven/src/main/java/org/openrewrite/maven/UpgradeDependencyVersion.java [hunk 2]
```diff
@@ -125,35 +126,79 @@ public String getDescription() {
...
                 Xml.Tag t = super.visitTag(tag, ctx);
                 try {
-                    if (isDependencyTag(groupId, artifactId)) {
+                    if (isPropertyTag()) {
+                        if (!tag.getValue().map(newVersion::equals).orElse(false)) {
+                            Path pomSourcePath = getResolutionResult().getPom().getRequested().getSourcePath();
+                            for (PomProperty pomProperty : accumulator.pomProperties) {
+                                if (pomProperty.pomFilePath.equals(pomSourcePath) &&
+                                    pomProperty.property.equals(tag.getName())) {
+                                    doAfterVisit(new ChangeTagValueVisitor<>(tag, newVersion));
```
Comment (timtebeek) [inline rewrite-maven/src/main/java/org/openrewrite/maven/UpgradeDependencyVersion.java:196]: Here we look to set the property value to `newVersion`, but `newVersion` can contain a [dependency version selector](https://docs.openrewrite.org/reference/dependency-version-selectors) pattern (e.g. `10.15.x`) if I'm not mistaken, as opposed to a literal version `10.15.0` as seen in the tests. We likely need to handle such cases as well, by determining the actual version value to fill in here. (unless I missed something, in which case please correct me!)

**Review by timtebeek — COMMENTED** [review]
Finally found a moment to go over your PR in more detail. Two questions arose as I went through the code. Would you be able to look into those?

File: rewrite-maven/src/main/java/org/openrewrite/maven/UpgradeDependencyVersion.java [hunk 3]
```diff
@@ -125,35 +126,79 @@ public String getDescription() {
...
+                        if (newerVersion != null) {
+                            Optional<Xml.Tag> version = tag.getChild("version");
+                            if (version.isPresent()) {
+                                String requestedVersion = d.getRequested().getVersion();
+                                if (requestedVersion != null && requestedVersion.startsWith("${") && !implicitlyDefinedVersionProperties.contains(requestedVersion)) {
+                                    String propertyName = requestedVersion.substring(2, requestedVersion.length() - 1);
+                                    if (!getResolutionResult().getPom().getRequested().getProperties().containsKey(propertyName)) {
+                                        getPomDeclaringProperty(getResolutionResult(), propertyName)
+                                                .ifPresent(pom -> accumulator.pomProperties.add(
+                                                        new PomProperty(requireNonNull(pom.getSourcePath()), propertyName)));
```
[the remark this answers is not in the document]
Comment (nguyenhoan) [reply rewrite-maven/src/main/java/org/openrewrite/maven/UpgradeDependencyVersion.java]: Make sense. Thanks for the suggestion!

File: rewrite-maven/src/main/java/org/openrewrite/maven/UpgradeDependencyVersion.java (hunk 2 above)
Comment (nguyenhoan) [reply rewrite-maven/src/main/java/org/openrewrite/maven/UpgradeDependencyVersion.java]: Excellent catch! Handled this in the next version.

File: rewrite-maven/src/main/java/org/openrewrite/maven/UpgradeDependencyVersion.java [hunk 4]
```diff
@@ -125,35 +126,81 @@ public String getDescription() {
...
+
+            @Override
+            public Xml.Tag visitTag(final Xml.Tag tag, final ExecutionContext ctx) {
+                ResolvedDependency d = findDependency(tag);
+                if (d != null && d.getRepository() != null) {
+                    // if the resolved dependency exists AND it does not represent an artifact that was parsed
+                    // as a source file, attempt to find a new version.
+                    try {
+                        String newerVersion = findNewerVersion(d.getVersion(),
+                                () -> downloadMetadata(d.getGroupId(), d.getArtifactId(), ctx), versionComparator, ctx);
```
Comment (timtebeek) [inline rewrite-maven/src/main/java/org/openrewrite/maven/UpgradeDependencyVersion.java:154]: Something I don't quite understand from looking over the code is that we appear to be looking at _any _dependency, not just dependencies whose groupId & artifactId match the one we're looking to upgrade. Am I reading that correctly?

File: rewrite-maven/src/main/java/org/openrewrite/maven/UpgradeDependencyVersion.java [hunk 5]
```diff
@@ -125,35 +126,81 @@ public String getDescription() {
...
+                accumulator.projectArtifacts.add(new GroupArtifact(pom.getGroupId(), pom.getArtifactId()));
+                return super.visitDocument(document, ctx);
+            }
+
+            @Override
+            public Xml.Tag visitTag(final Xml.Tag tag, final ExecutionContext ctx) {
+                ResolvedDependency d = findDependency(tag);
+                if (d != null && d.getRepository() != null) {
+                    // if the resolved dependency exists AND it does not represent an artifact that was parsed
+                    // as a source file, attempt to find a new version.
```
Comment (timtebeek) [inline rewrite-maven/src/main/java/org/openrewrite/maven/UpgradeDependencyVersion.java:151]:
```suggestion
                if (isDependencyTag(groupId, artifactId)) {
                    // if the resolved dependency exists AND it does not represent an artifact that was parsed
                    // as a source file, attempt to find a new version.
                    ResolvedDependency d = findDependency(tag);
```

**Review by timtebeek — APPROVED** [review]
Thanks so much for getting this started @nguyenhoan , and your patience as I had to wrap my head around the changes. I've added a few smaller changes, mostly to familiarize myself with the code, and to better accommodate changes I expect to work in next. Great addition, and neat to better support such minimal property bumps from now on.

**Comment by nguyenhoan** [comment]
Thank you for helping with this important enhancement and bringing it to the destination!

Could you please let me know when the relevant changes to other modules such as `org.openrewrite.java.dependencies.UpgradeDependencyVersion` in [rewrite-java-dependencies](https://github.com/openrewrite/rewrite-java-dependencies) are also merged and the new releases in tandem happen? Thanks!

**Comment by timtebeek** [comment]
I've just opened a PR for rewrite-java-dependencies
- https://github.com/openrewrite/rewrite-java-dependencies/pull/103

We release snapshot versions every day, and new releases of all openrewrite recipe modules typically every two weeks. We previous release was last week; so then likely by next week there should be a release containing this improvement.

**Comment by nguyenhoan** [comment]
Awesome! Thanks!

**Comment by nguyenhoan** [comment]
Yay! Thank you!

**Pull request merged**